### PR TITLE
topic: Introduce Test Doubles

### PR DESCRIPTION
**Pull request by EdDev**

### CODE REVIEW

File: topics/isthisamock.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+```go
+// Same production code as before
+
+// Test Stub
+type MockMonitor struct {
+    displayData         string
+    expectedDisplayData string
+}
+func (m *MockMonitor) Display(fmt string, args ...any) {
+    m.data = fmt.Sprintf(fmt, args...)
```
Comment (jpeimer) [inline topics/isthisamock.md]:
```suggestion
    m.displayData = fmt.Sprintf(fmt, args...)
```

File: topics/isthisamock.md [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+type StubMonitorDisplayFails struct {
+    err error
+}
+func (s StubMonitorDisplayFails) Display(fmt string, args ...any) {
+	return s.err 
+}
+
+// Test
+func testComputerIsStatringWithDisplayFailure() {
+    monitor := StubMonitorDisplayFails{errors.New("an error")}
```
Comment (jpeimer) [inline topics/isthisamock.md:105]: Should be this? (I haven't written in Go that much, may just not understand)
```suggestion
    monitor := StubMonitorDisplayFails{err.New("an error")}
```

File: topics/isthisamock.md [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+Mocks have become popular because tooling have been built to generate
+them for a given interface automatically.
+
+Its disadvantage is similar to spies.
+It requires the setup stage and mixes expectations with assertions and
+when used with autogenerated tooling, it usually does a lot more than it
+is needed.
+
+## Fake
```
Comment (EdDev) [inline topics/isthisamock.md:185]: Here is a real life example of a fake: https://github.com/kubevirt/kubevirt/blob/c6b4ce6ea8c3ae08b4ffb83b815b7727e854477e/pkg/network/driver/netlink/fake/fake.go#L28

It mimics this code: https://github.com/kubevirt/kubevirt/tree/c6b4ce6ea8c3ae08b4ffb83b815b7727e854477e/pkg/network/driver/netlink

File: topics/isthisamock.md [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        return err
+    }
+	// retry 10 times to verify that the service is running
+    for i := 0, i < 10, i++ {
+        if status := service.Status(); status == "running" {
+            return nil
+		}
+        time.Sleep(1 * time.Milisecond)
+    }
+    return errors.New("unable to run service, timeout reached")
```
Comment (jpeimer) [inline topics/isthisamock.md:211]: Same question here 
```suggestion
    return err.New("unable to run service, timeout reached")
```

File: topics/isthisamock.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    return 0
+}
+
+// Test Dummy
+type DummyMonitor struct {}
+func (_ DummyLog) Display(fmt string, args ...any) { return nil }
+
+// Test
+func testNewComputerIsNotRunning() {
+    c := NewComputer(DummyMonitor{})
```
Comment (jpeimer) [inline topics/isthisamock.md]: Would suggest giving them the full names, but up to you :) 
```suggestion
    computer := NewComputer(DummyMonitor{})
```

**Review by jpeimer — CHANGES_REQUESTED** [review]

File: topics/isthisamock.md (hunk 5 above)
Comment (EdDev) [reply topics/isthisamock.md]: Usually for vars that are defined and used very closely, it is perfectly fine to give them short names.

But I change it as it seems to confuse some readers.
Done.

File: topics/isthisamock.md (hunk 2 above)
Comment (EdDev) [reply topics/isthisamock.md:105]: This is the [errors](https://pkg.go.dev/errors#New) package.
So this is fine.

File: topics/isthisamock.md (hunk 1 above)
Comment (EdDev) [reply topics/isthisamock.md]: Done

File: topics/isthisamock.md (hunk 4 above)
Comment (EdDev) [reply topics/isthisamock.md:211]: This is the errors package.
So this is fine.

**Comment by EdDev** [comment]
@jpeimer , thank you for the review!

**Comment by EdDev** [comment]
[change](https://github.com/wannaknowmore/codequality/compare/25ca87411eef629fa1ad02ecccf04bc9434a2eb3..0ee20c6e805b1757b70e02a1663b8c71abcd9e35): Responded to review.

**Review by jpeimer — APPROVED** [review]

**Pull request merged**